### PR TITLE
Reasoning parser

### PR DESCRIPTION
**Pull request by xihuai18**
Based on #3202 and #3859, I fixed some bugs, simplified the implementations and added new features.

1. Only support `--reasoning-parser` server argument now. If users want to separate reasoning content, they must specify this option. **The parser is chosen only based on this option, but not the model name.**
2. `separate_reasoning` and `stream_reasoning` are set as `True` by default.
3. Add support for SGLang native api and offline engine.
4. Add and `.ipynb` docs.
5. Refactor and fix some bugs.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [x] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by xihuai18** [comment]
@shuaills @zhaochenyang20 @tot0 @zhyncs  I think this PR is ready for merging now, could you review?

**Comment by zhaochenyang20** [comment]
@shuaills could you please take a look.

**Review by shuaills — APPROVED** [review]

File: python/sglang/srt/reasoning_parser.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
+import re
+from typing import Dict
+
+REASONING_MODELS = ["deepseek-r1"]
```
Comment (lupickup) [inline python/sglang/srt/reasoning_parser.py]: I think it's better to not duplicate the list of reasoning models. Removing `is_reasoning_model()` makes sense, I think it's cleaner to get the list of supported models from the `ReasoningParser` class though.
```suggestion
```

File: python/sglang/srt/server_args.py [hunk 2]
```diff
@@ -23,6 +23,7 @@
 import torch
 
 from sglang.srt.hf_transformers_utils import check_gguf_file
+from sglang.srt.reasoning_parser import REASONING_MODELS
```
Comment (lupickup) [inline python/sglang/srt/server_args.py]:
```suggestion
from sglang.srt.reasoning_parser import ReasoningParser
```

File: python/sglang/srt/server_args.py [hunk 3]
```diff
@@ -614,6 +616,13 @@ def add_cli_args(parser: argparse.ArgumentParser):
             action="store_true",
             help="Return number of cached tokens in usage.prompt_tokens_details for each openai request.",
         )
+        parser.add_argument(
+            "--reasoning-parser",
+            type=str,
+            choices=REASONING_MODELS,
```
Comment (lupickup) [inline python/sglang/srt/server_args.py]:
```suggestion
            choices=ReasoningParser.DetectorMap.keys(),
```

File: python/sglang/srt/server_args.py [hunk 4]
```diff
@@ -614,6 +616,13 @@ def add_cli_args(parser: argparse.ArgumentParser):
...
+            default=ServerArgs.reasoning_parser,
+            help=f"Specify the parser for reasoning models, supported parsers are: {REASONING_MODELS}.",
```
Comment (lupickup) [inline python/sglang/srt/server_args.py]:
```suggestion
            help=f"Specify the parser for reasoning models, supported parsers are: {ReasoningParser.DetectorMap.keys()}.",
```

File: python/sglang/srt/server_args.py [hunk 5]
```diff
@@ -96,6 +97,7 @@ class ServerArgs:
     api_key: Optional[str] = None
     file_storage_pth: str = "sglang_storage"
     enable_cache_report: bool = False
+    reasoning_parser: Optional[str] = None
```
Comment (lupickup) [inline python/sglang/srt/server_args.py:101]: To remain compatible with clients which haven't implemented support for DeepSeeks default behavior (`separate_reasoning=True`) Microsoft is serving DeepSeek-R1 with `separate_reasoning=False` as the default.

Can we add an argument to change the default for `separate_reasoning`? Or an environment variable? Either way this is crucial functionality for us.

**Review by lupickup — CHANGES_REQUESTED** [review]

File: python/sglang/srt/reasoning_parser.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            raise ValueError(f"Unsupported model type: {model_type}")
+
+        self.detector = detector_class(stream_reasoning=stream_reasoning)
+
+    def parse_non_stream(self, full_text: str) -> StreamingParseResult:
+        """Non-streaming call: one-time parsing"""
+        ret = self.detector.detect_and_parse(full_text)
+        return ret.reasoning_text, ret.normal_text
+
+    def parse_stream_chunk(self, chunk_text: str) -> StreamingParseResult:
```
Comment (lupickup) [inline python/sglang/srt/reasoning_parser.py]: This is no longer the return type of this method:
```suggestion
    def parse_stream_chunk(self, chunk_text: str) -> Tuple[str, str]:
```

File: python/sglang/srt/reasoning_parser.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        if not model_type:
+            raise ValueError("Model type must be specified")
+
+        detector_class = self.DetectorMap.get(model_type.lower())
+        if not detector_class:
+            raise ValueError(f"Unsupported model type: {model_type}")
+
+        self.detector = detector_class(stream_reasoning=stream_reasoning)
+
+    def parse_non_stream(self, full_text: str) -> StreamingParseResult:
```
Comment (lupickup) [inline python/sglang/srt/reasoning_parser.py]:
```suggestion
    def parse_non_stream(self, full_text: str) -> Tuple[str, str]:
```

File: python/sglang/srt/reasoning_parser.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
+import re
+from typing import Dict
```
Comment (lupickup) [inline python/sglang/srt/reasoning_parser.py]:
```suggestion
from typing import Dict, Tuple
```

**Review by lupickup — CHANGES_REQUESTED** [review]
Fix up type hints in `reasoning_parser.py`

File: python/sglang/srt/reasoning_parser.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
+import re
+from typing import Dict, Tuple
+
+
```
Comment (lupickup) [inline python/sglang/srt/reasoning_parser.py:4]: There's one extra line here because I wasn't able to select multiple lines for suggestion :)
```suggestion
```

File: python/sglang/srt/server_args.py (hunk 5 above)
Comment (lupickup) [reply python/sglang/srt/server_args.py:101]: Changes proposed in #4001

File: python/sglang/srt/server_args.py [hunk 10]
```diff
@@ -619,9 +619,9 @@ def add_cli_args(parser: argparse.ArgumentParser):
         parser.add_argument(
             "--reasoning-parser",
             type=str,
-            choices=ReasoningParser.DetectorMap.keys(),
+            choices=list(ReasoningParser.DetectorMap.keys()),
```
Comment (lupickup) [inline python/sglang/srt/server_args.py:639]: Nice catch

File: python/sglang/srt/entrypoints/http_server.py [hunk 11]
```diff
@@ -394,6 +396,26 @@ async def function_call_request(obj: ParseFunctionCallReq, request: Request):
...
+    # 2) Call the non-stream parsing method (non-stream)
+    reasoning_text, normal_text = parser.parse_non_stream(obj.text)
+
+    # 3) Organize the response content
+    response_data = {
+        "reasoning_text": reasoning_text,
+        "text": normal_text,
+    }
+
+    return ORJSONResponse(content=response_data, status_code=200)
```
Comment (shuaills) [inline python/sglang/srt/entrypoints/http_server.py:482]: Can we have a non-streaming endpoint?

File: docs/backend/separate_reasoning.ipynb [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+    "```"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": []
+  }
```
Comment (shuaills) [inline docs/backend/separate_reasoning.ipynb]: Remove this empty block.

**Review by shuaills — CHANGES_REQUESTED** [review]
Great Job!

File: python/sglang/srt/openai_api/adapter.py [hunk 13]
```diff
@@ -1099,9 +1105,24 @@ def v1_chat_generate_response(
...
             tools = request[idx].tools
+            separate_reasoning = request[idx].separate_reasoning
         else:
             tool_choice = request.tool_choice
             tools = request.tools
+            separate_reasoning = request.separate_reasoning
+
+        if reasoning_parser and separate_reasoning:
+            try:
+                parser = ReasoningParser(reasoning_parser, True)
```
Comment (XiaotongJiang) [inline python/sglang/srt/openai_api/adapter.py]: can we use named assignment for readability

File: python/sglang/srt/reasoning_parser.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+        # https://github.com/sgl-project/sglang/pull/3202#discussion_r1950153599
+
+
+class ReasoningParser:
+    """
+    Parser that handles both streaming and non-streaming scenarios for extracting
+    reasoning content from model outputs.
+
+    Args:
+        model_type (str): Type of model to parse reasoning from
```
Comment (XiaotongJiang) [inline python/sglang/srt/reasoning_parser.py:127]: nit: what is model type and whats the reason to not just use model name?

File: test/srt/test_reasoning_content.py [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+            "max_tokens": 100,
+            "stream": True,
+            "extra_body": {"separate_reasoning": True, "stream_reasoning": False},
+        }
+        response = client.chat.completions.create(**payload)
+
+        reasoning_content = ""
+        content = ""
+        first_chunk = False
+        for chunk in response:
```
Comment (XiaotongJiang) [inline test/srt/test_reasoning_content.py:125]: can we assert chunk size > 1 to make sure theres more than one chunk?

File: python/sglang/srt/reasoning_parser.py (hunk 14 above)
Comment (lupickup) [reply python/sglang/srt/reasoning_parser.py:127]: The docs actually refer to this as "model series", it's not quite equivalent to model_name because parser likely apply to a set of models which use the same reasoning tokens. The "deepseek-r1" parser applies to all the R1 distill models (e.g. "deepseek-ai/deepseek-r1-distill-llama-8b").

**Comment by lupickup** [comment]
So is the consensus to not include this change? https://github.com/sgl-project/sglang/commit/be0317144589fe0b486906294efb65e9dc2c60af

I'd like to understand the reason why if so.

**Comment by zhaochenyang20** [comment]
> So is the consensus to not include this change? [be03171](https://github.com/sgl-project/sglang/commit/be0317144589fe0b486906294efb65e9dc2c60af)
> 
> I'd like to understand the reason why if so.

cc @shuaills @XiaotongJiang

File: python/sglang/srt/entrypoints/http_server.py (hunk 11 above)
Comment (xihuai18) [reply python/sglang/srt/entrypoints/http_server.py:482]: it is non streaming

File: python/sglang/srt/reasoning_parser.py (hunk 14 above)
Comment (xihuai18) [reply python/sglang/srt/reasoning_parser.py:127]: user can call any model name they want, use model name for distinction is not a good choice

File: test/srt/test_reasoning_content.py (hunk 15 above)
Comment (xihuai18) [reply test/srt/test_reasoning_content.py:125]: for streaming API, it must have >2 chunks. The real first chunk indicates the role, the second one begins the content.

File: python/sglang/srt/openai_api/adapter.py (hunk 13 above)
Comment (xihuai18) [reply python/sglang/srt/openai_api/adapter.py]: fixed

**Comment by xihuai18** [comment]
Do we need more things to merge this PR ?

**Comment by zhaochenyang20** [comment]
cc @zhyncs could you take a look?

**Review by lupickup — APPROVED** [review]
